working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,290 @@
+     literals, in source order.
+
+   * Horizontal inference: if this is not the last stage, partially solve the
+     set of type constraints accumulated so far, to produce an updated
+     preliminary mapping of type parameters to type schemas.  Otherwise proceed
+     to step 5.
+
+5. Upwards inference: solve the set of type constraints accumulated so far, to
+   produce a final mapping of type parameters to types.  Check that each type is
+   a subtype of the bound of its corresponding type parameter.
+
+6. Type checking: Check that the static type of each argument is assignable to
+   the type obtained by substituting the final mapping (from step 5) into the
+   corresponding parameter type of the invocation target.
+
+7. Assign static type: Finally, obtain the static type of the invocation by
+   substituting the final mapping (from step 5) into the return type of the
+   invocation target.
+
+(Note that step 4 is the only step that differs from the inference implemented
+in Dart today, and it only takes effect if the experiment flag
+`inference-update-1` is enabled, hence this feature is backward compatible.)
+
+### Topological sort
+
+In the algorithm above, the "topological sort" in step 4 works as follows.
+First there is a _dependency analysis_ phase, in which type inference decides
+which invocation arguments might benefit from being type inferred before other
+arguments, and then a _stage selection_ phase, in which type inference groups
+the arguments into stages.
+
+#### Dependency analysis
+
+First, we form a dependency graph among the invocation arguments based on the
+following rule: there is a dependency edge from argument _A_ to argument _B_ if
+and only if the type of the invocation target is generic, and the following
+relationship exists among _A_, _B_, and at least one of the invocation target’s
+type formals _T_:
+
+1. _A_ is a function literal (see [what constitutes a function
+   literal?](#what-constitutes-a-function-literal))
+
+2. AND the parameter in the invocation target corresponding to _A_ is function
+   typed,
+
+3. AND _T_ is a free variable in at least one of the parameters of that function
+   type,
+
+4. AND the corresponding parameter in _A_ is implicitly typed,
+
+5. AND EITHER:
+
+   * The parameter in the invocation target corresponding to _B_ is function
+     typed, and _T_ is a free variable in its return type
+
+   * OR the parameter in the invocation target corresponding to _B_ is _not_
+     function typed, and _T_ is a free variable in its type.
+
+(The idea here is that we're trying to draw a dependency edge from _A_ to _B_ in
+precisely those cirumstances in which there is likely to be a benefit to
+performing a round of horizontal inference between type inferring _A_ and type
+inferring _B_).
+
+If the type of the invocation target is `dynamic` or the `Function` class, or
+some non-generic function type, then the resulting graph has no edges.
```

Comment:
Fair point.  However, since there is no existing specification for type inference of invocations, and I'm trying to supply one as part of this proposal, it seemed reasonable to include details about what happens in the case of dynamic invocations.

Reply:
Thanks!